### DISCUSSION TITLE
Modbus, Macro Buttons, etc.

### DISCUSSION THREAD

**Discussion by mintracer** [grblHAL/core#544] [Ideas]
Hei ;)

last weekend i hooked up a modbus relay board to my grblhal2k-EST, to control it i use the "MBIO" Plugin from Risototh - 

https://github.com/risototh/grblHAL-modbus-io

had some issues to compile it since there were missing some semicolon (label and decleration order problem) and a error message problem with a "Status_ExpressionInvalidResult", so everytime switching a reset occured... i handeled that by changing the error message type to "Status_OK" quite pragmatic :D

So now to my wish or idea...

would it be possible to make the macro buttons and / or m-codes the same priority as the realtime buttons, such as start, stop, hold and so on? 

This would be awesome when using a button panel to switch on an off some thing during a running job

have a nice week! ;)

cheers Einar! :)

**Comment by karoria** [comment]
M64 and M65 are used for turning on and turning off (respectively) digital outputs immediately. That should do your task. E.g. M64P1 makes aux output 1 on immediately. This command skips planner buffer for execution. So it executes immediately when read.
Cheers.
Ravi

**Reply by mintracer** [reply]
but does M64 add also Modbus Ports? what if when i use a 32 Channel Relay board? Would M64P28 also work?

**Comment by terjeio** [comment]
This can be done but not via the keypad macro plugin since that can only execute macros when in grblHAL is in idle state. A new plugin that claims aux inport ports and add interrupt handlers that calls functions such as `mbio_ModBus_WriteRegister()` in the mbio plugin on button events has to be added. Note the the interrupt handlers has to post a handler function/data to the foreground process (via `task_add_immediate()`) since functions such as `mbio_ModBus_WriteRegister()` are not safe to call from an interrupt context.

E.g. the [template macro plugin](https://github.com/grblHAL/Templates/blob/master/my_plugin/Macros_bound_to_aux_inputs/my_plugin.c) can be modified for this by changing the following functions to something like this:

```
#define DEVICE_ADDRESS 1

static void run_macro (void *data)
{
    mbio_ModBus_WriteRegister(DEVICE_ADDRESS, (uint16_t)data, 1);
}

ISR_CODE static void execute_macro (uint8_t irq_port, bool is_high)
{
    uint_fast8_t idx = N_MACROS;
    do {
        idx--;
    } while(idx && port[idx] != irq_port);

    task_add_immediate(run_macro, (void *)idx);
}
```
Buttons could be used to toggle the outputs by storing the previous state and inverting it on each interrupt.
